iroh-gateway/src/core.rs

```diff
@@ -0,0 +1,412 @@
+use axum::{
+    body::{self, Body, BoxBody},
+    error_handling::HandleErrorLayer,
+    extract::{Extension, Path, Query},
+    http::{header::*, StatusCode},
+    response::IntoResponse,
+    routing::get,
+    BoxError, Router,
+};
+use serde::Deserialize;
+use std::{borrow::Cow, collections::HashMap, time::Duration};
+use tower::ServiceBuilder;
+
+use crate::{
+    client::{Client, Request},
+    config::Config,
+    constants::*,
+    error::GatewayError,
+    response::{GatewayResponse, ResponseFormat},
+};
+
+pub struct Core {
+    pub config: Config,
+    client: Client,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetParams {
+    format: Option<String>,
+    filename: Option<String>,
+    download: Option<String>,
+}
+
+impl Core {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            client: Client::new(),
+        }
+    }
+
+    pub async fn serve(&self) {
+        let app = Router::new()
+            .route("/ipfs/:cid", get(get_ipfs))
+            .route("/ipfs/:cid/*cpath", get(get_ipfs))
+            .layer(Extension(self.config.clone()))
```

Comment:
the config and client should probably be wrapped into an `Arc` so that cloning is cheap and not the full struct gets cloned, but rather a reference